mlir-sycl/doc/design/SYCLKernelSubmissionRepresentation.md

```diff
@@ -0,0 +1,510 @@
+the integration header. In particular, this means that accessors are "passed" as
+a pointer to an instance of one of the `sycl::accessor` classes, rather than
+being broken down into their parts. Accessor- and local accessor-arguments are
+annotated with their respective SYCL dialect type to preserve high-level
+information such as the access mode. Other *standard layout* objects such as
+generic structs and scalars are captured after reverting any copy optimizations
+done by the frontend (see [below](#captured-values)).
+
+At this point, the handler operand is merely an implementation detail which
+prevents the operation from being considered trivially dead without interfering
+with the dataflow analyses defined in context of the SYCL dialect. Also note
+that we do not model events or dependencies between different kernel launches
+yet — a future extension of the operation may have a result and operands
+representing events.
+
+## Raising approach
+
+Lots of high-level structure and information is lost during the lowering of the
+object-oriented SYCL source code to LLVM-IR. In order to recover the information
+required for the `schedule_kernel` operation, we employ a multi-stage pattern
+matching approach that is aided by source code annotations in the SYCL headers.
+The following subsections describe the intermediate steps in detail. The
+implementation resides in
+`polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp`. Note that
+example MLIR snippets have been manually edited to use descriptive names for
+values and other identifiers.
+
+### Command-group handler
+
+All raising patterns operate on the CGF itself or annotation known to reside in
+the CGF after mandatory inlining (see
+[above](#prerequisites--known-limitations)). Hence, the handler is first
+argument (after the `this` pointer) of the CGF.
+
+### Host constructors
+
+We raise calls/invokes to specific constructors of SYCL API classes (`buffer`,
+`accessor`, `range`, etc.) to the
+[`sycl.host.constructor`](#syclhostconstructor-syclsyclhostconstructorop) op,
+which ties together arguments for construction and the SYCL dialect type of the
+entity being initialized.
+
+For example, the instantiation of an accessor in the CGF results in the
+following raised IR:
+```mlir
+%accessor = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::accessor", // ...
+...
+sycl.host.constructor(%accessor, %buffer, %handler, %propertylist, %codeloc)
+  {type = !sycl_accessor_1_21llvm2Evoid_r_gb}
+    : (!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> ()
+```
+
+Note that in contrast to the constructor operations
+intended for the modeling of device code, the host-side constructor operation
+does *not* include the allocation of the object. Hence, a common pattern in the
+host raising process is looking for a `host.constructor` op in the users of a
+particular `llvm.alloca` operation.
+
+> TODO: Elaborate here, or in a separate document?
```

Comment:
I think we can elaborate here and keep related information in one place.

Reply:
Yes, maybe it would be good to mention the type attribute, how it does not correspond to the original one in the code and how we are encoding some additional information in that type